library/core/src/init.rs

```diff
@@ -0,0 +1,86 @@
+//! In-place initialization.
+//!
+//! This module describes the interface through which types supporting in-place initialization
+//! can perform initialization with minimal or zero additional allocations or moves.
+
+use crate::ptr::Pointee;
+
+/// In-place Initializer for `T`.
+///
+/// An instance of `Init<T>` carries all the information necessary to initialize a `T` in a
+/// well-defined memory location, criteria of which is prescribed in the Safety section.
+///
+/// # Fallibility
+///
+/// The initialization might fail and return an error of type [`Self::Error`] instead.
+/// In that case, the memory provided to [`Self::init`] shall be repurposed in any way,
+/// even though it might have been written to by this initializer.
+///
+/// # Examples
+///
+/// ## Initializing unsized types
+///
+/// To initialize an unsized type, first query the required layout for `T` using [`Self::layout`].
+/// Then provide a pointer to an allocation of at least the specified alignment and size.
+///
+/// If initialization was successful, then [`Self::init`] returns the metadata that combined with
+/// the pointer to the given to [`Self::init`] yields a valid pointer to `T`.
+///
+/// ```
+/// use std::alloc::alloc;
+/// fn init_unsized<T: ?Sized + Pointee, I: Init<T>>(init: I) -> Result<Box<T>, I::Error> {
```

Comment:
Could you show how this gets called as well?

It would be good to copy the example to the PR description too to give everyone a quick overview.

Reply:
Yeah, let's.